### DISCUSSION TITLE
What is the correct use, if any, of AsyncResult.then()?

### DISCUSSION THREAD

**Discussion by MajorDallas** [celery/celery#9201] [Q&A]
I spotted the `.then` method in the [API reference](https://docs.celeryq.dev/en/stable/reference/celery.result.html#celery.result.ResultSet.then), read the code a bit and learned that this method is wrapping the same-named method on a [vine promise](https://github.com/celery/vine/blob/master/vine/promises.py#L184). Other than being listed in the API reference, there is no documentation about this method at all.

My guess was that it would have the same overall effect as the `link` parameter to `apply_async()`, if in a different way (i.e. not as a separate task in chain but as a plain old function to run in the worker once the task completes). Given the sparse documentation on the method, I tried some simple experiments with a lambda function to see if this was the case: first, it just returned all the arguments it received; second, it printed them; third, it logged them. The effect of this never materialized, however. Either I'm invoking it wrong, or my guess was off.

My next guess would be that the `then` method is meant for internal use and I shouldn't use it. That's fine if so, I just want to confirm.

**Comment by thedrow** [comment]
Do you have a code example I can run to investigate?
If `then` is public API, it should work as declared.

**Reply by MajorDallas** [reply]
I'd already deleted the sandbox code where I tested the first time, but was able to write a couple very simple tasks to test with:

```python
@app.task
def first(x):
    return x + 1

@app.task
def second(y):
    return y * 2

# and in a repl:

first.s(1).apply_async().then(second.s())
# <promise@0x70128b585ea0 --> mymodule.second>
first.s(1).apply_async().then(second.s())()  # or first.s(1).apply_async().then(second)()
TypeError: second() missing 1 required positional argument: 'y'
first.s(1).apply_async().then(lambda r: print(r))
# <promise@0x70128ae44400 --> <function <lambda> at 0x70128b197820>>
first.s(1).apply_async().then(lambda r: print(r))()
# TypeError: <lambda>() missing 1 required positional argument: 'r'
r = first.s(1).apply_async(link=(second.s(),))
r.get()  # see note below
# 4
first.s(1).apply_async(link=(second.s(),)).then(lambda *r: print(r))()
# ()
```

So I learned in this new sandbox that a `promise` must be called before it does anything. That helps me understand _how_ to use it, but I'm still uncertain about _when_.

Note: Unrelated to the main question, but a heads-up: Somehow this method of linking the tasks together results in a TypeError in the repl, claiming `AsyncResult` and `int` can't be multiplied; meanwhile, the Celery worker's logs show it multiplying the two expected integers and returning `4`. It doesn't occur if I use `chain`, so I'm sure I'm just calling it wrong, but I wanted to mention it in case you find the same thing.

**Comment by shahabRDZ** [comment]
I went down this rabbit hole too. `.then()` is basically a leftover from the vine promise system and it is... barely functional in practice.

from what I have gathered:

- `.then()` registers a callback on the vine promise object
- it only fires if something actively resolves the promise, which in celery means polling or getting the result
- it does NOT fire automatically when the task completes on the worker side
- it only works on the client side, not the worker side

so this will NOT do what you expect:

```python
result = my_task.delay()
result.then(lambda r: print(f"done: {r}"))
# nothing happens unless you later call result.get() or result.result
```

and this WILL work but is kind of pointless:

```python
result = my_task.delay()
result.then(lambda r: print(f"done: {r}"))
result.get()  # blocks, then the callback fires
# at this point... you already have the result, why use .then()?
```

**what you actually want:**

- for chaining work after a task completes → use `link`:
  ```python
  my_task.apply_async(args=[...], link=on_complete.s())
  ```

- for fire-and-forget notifications → use signals:
  ```python
  from celery.signals import task_success
  
  @task_success.connect(sender=my_task)
  def on_success(result, **kwargs):
      print(f"done: {result}")
  ```

- for monitoring from the client side → poll with `AsyncResult.ready()`

honestly `.then()` should either be properly documented or deprecated. right now it is just a trap.

**Reply by auvipy** [reply]
are you up for the contribution?

**Reply by MajorDallas** [reply]
What I was trying to accomplish actually forced me into a position where I can't know if I need to link a task until after it's already been started. In distilled form:

```python
@task
def provision_vm(**kwargs):
    ...


def reserve_vm_for_user(user, **kwargs):
    # We have a pool of VMs, and we know one way or another we need to provision a VM, so we start immediately
    # so that the user waits for the shortest possible time if the pool happens to be empty.
    first = provision_vm.s(**kwargs).apply_async()
    # Now we check the pool to see if there's one we don't have to wait for. Each one has to be checked, so this can
    # be pretty slow, which is why we start the first provision as soon as possible.
    readies = find_ready_vms(**kwargs)
    if not readies:
        first.then(notify_user)  # This will be the user's new VM, and we'll start another for the pool.
        second = provision_vm.s(**kwargs).apply_async(link=add_to_pool.s())
    else:
        notify_user(readies[0])
        first.then(lambda vm: add_to_pool.s(vm).apply_async())
    ThreadPoolExecutor(1).submit(first.get)  # how I'm working around having to block on `.get()`
    return 200
```

This is an early version of a new feature: the hope is to manage the pool well enough that this code never has to worry about not getting a ready VM.

Although, I suppose a better approach might be to set a flag in a db or somewhere and define a new callback task that checks that flag and branches there, rather than that branch happening in the task caller :thinking: Kinda wish I'd thought of that 9 months ago...
